triton/writer_batch.go

```diff
@@ -0,0 +1,148 @@
+package triton
+
+import "time"
+
+const defaultBatchSize = 10
+const defaultFlushInterval = 1 * time.Second
+
+// MaxBatchSize is the limit Kinesis has on a PutRecords call
+const MaxBatchSize = 500
+
+// NewBatchWriter creates a batch version of an existing Writer using default
+// values for size and interval.
+func NewBatchWriter(w Writer) *BatchWriter {
+	return NewBatchWriterSize(w, defaultBatchSize, defaultFlushInterval)
+}
+
+// NewBatchWriterSize creates a batch writer using the given size and interval.
+func NewBatchWriterSize(w Writer, size int, intr time.Duration) *BatchWriter {
+	if size <= 0 {
+		size = defaultBatchSize
+	} else if size > MaxBatchSize {
+		size = MaxBatchSize
+	}
+
+	if intr <= 0 {
+		intr = defaultFlushInterval
+	}
+
+	bufferSize := 10000
+
+	bw := &BatchWriter{
+		w:      w,
+		buf:    make(chan Record, bufferSize),
+		signal: make(chan struct{}, bufferSize),
+		errors: make(chan error, 1),
+
+		size: size,
+		intr: intr,
+
+		ticker: time.NewTicker(intr),
+	}
+
+	go bw.writeLoop()
+
+	return bw
+}
+
+// BatchWriter implements an asyncronous writer that writes records in batches.
+// A batch is written when either the buffer size is exceeded or the time
+// interval since the last write has been exceeded.
+//
+// Write errors are written to the Errors() channel. It is highly recommended
+// that you actively monitor this channel because the writer will not stop
+// after an error.
+type BatchWriter struct {
+	w      Writer
+	buf    chan Record
+	errors chan error
+
+	size int
+	intr time.Duration
+
+	ticker *time.Ticker
+	signal chan struct{}
+}
+
+func (bw *BatchWriter) writeLoop() {
+	for {
+		select {
+		case _, ok := <-bw.signal:
+			if !ok {
+				return
+			}
+
+			if len(bw.buf) >= bw.size {
+				bw.flush()
+			}
+		case <-bw.ticker.C:
+			bw.flush()
+		}
+	}
+}
+
+// flush writes everything in the current buffer
+func (bw *BatchWriter) flush() {
+	numBuffered := len(bw.buf)
+	if numBuffered == 0 {
+		return
+	}
+
+	records := make([](Record), len(bw.buf))
+	for i := 0; i < numBuffered; i++ {
+		records[i] = <-bw.buf // this will never block
+	}
+
+	// write in blocks of batch size
+	for n := 0; n < numBuffered; n += bw.size {
+		end := n + bw.size
+		if end > numBuffered {
+			end = numBuffered
+		}
+
+		if err := bw.w.WriteRecords(records[n:end]...); err != nil {
+			bw.writeError(err)
+		}
+	}
+}
+
+// Flush forces all buffered records to be sent.
+// If there is an error it will have been written to the Errors chan.
+func (bw *BatchWriter) Flush() {
```

Comment:
I have mixed feeling about whether this should return an error or delegate to the Errors() channel. Errors channel seems cleaner, but means you would need to check it any time you Flush or Close. Thoughts?